l10n_th_personal_income_tax/models/pit_rate.py

```diff
@@ -0,0 +1,139 @@
+        for rec in self:
+            prev_income_to = 0.0
+            accum_tax = 0.0
+            for i, rate in enumerate(rec.rate_ids):
+                if i == 0 and rate.income_from != 0.0:
+                    raise UserError(_("Income amount must start from 0.0"))
+                if i > 0 and float_compare(rate.income_from, prev_income_to, 2) != 0:
+                    raise UserError(
+                        _(
+                            "Discontinued income range!\n"
+                            "Please make sure Income From = Previous Income To"
+                        )
+                    )
+                prev_income_to = rate.income_to
+                accum_tax += rate.amount_tax_max
+                rate.amount_tax_max_accum = accum_tax
+
+    def _compute_total_wt(self, rate_amount, tax_rate):
+        return rate_amount * (tax_rate / 100)
+
+    @api.model
+    def calculate_rate_wt(self, total_income, income, date=False):
+        if not date:
+            date = fields.Date.context_today(self)
+        calendar_year = date.strftime("%Y")
+        pit = self.search(
+            [("effective_date", "<=", date)], order="effective_date desc", limit=1
+        )
+        if not pit:
+            raise ValidationError(
+                _("No PIT Rate Table found for calendar year %s") % (calendar_year,)
+            )
+        rate_ranges = pit.rate_ids.filtered(lambda l: abs(total_income) > l.income_from)
+        current_amount = 0.0
+        income_residual = income
+        total_income_residual = total_income
+        for rate_range in reversed(rate_ranges):
+            rate_amount = total_income_residual - rate_range.income_from
+            if income > rate_amount and income_residual >= rate_amount:
+                total_income_residual -= rate_amount
+                income_residual -= rate_amount
+                current_amount += self._compute_total_wt(
+                    rate_amount, rate_range.tax_rate
+                )
+            else:
+                current_amount += self._compute_total_wt(
+                    income_residual, rate_range.tax_rate
+                )
+                break
+        return current_amount
+
+
+class PersonalIncomeTaxRate(models.Model):
+    _name = "personal.income.tax.rate"
+    _description = "PIT Rates"
+    _order = "sequence, id"
+
+    sequence = fields.Integer(string="sequence")
+    pit_id = fields.Many2one(
+        comodel_name="personal.income.tax",
+        string="PIT Table",
+        ondelete="cascade",
+        index=True,
+    )
+    income_from = fields.Float(
+        string="Income From (>)",
+    )
+    income_to = fields.Float(
+        string="Income To (<=)",
+    )
+    tax_rate = fields.Float(
+        string="Tax Rate (%)",
+    )
+    amount_tax_max = fields.Float(
+        string="Maximum Tax in Range",
+        compute="_compute_amount_tax_max",
+        store=True,
+    )
+    amount_tax_max_accum = fields.Float(
```

Comment:
I think this can be computed, non-stored field. You just want to display it for user? (or do we really need this field?)